### DISCUSSION TITLE
Doing a transaction procedurally, instead of passing in a closure?

### DISCUSSION THREAD

**Discussion by vdods** [diesel-rs/diesel#3214] [Q&A]
I've been using Connection::transaction, to which a closure is passed which comprises the body of the transaction.  I'm finding it very hard to compose this with other functions that invoke DB statements.  Is there a way to do this more like in plain SQL, where you issue a begin transaction statement, and later issue a commit statement?  Perhaps the begin transaction would return a guard object that rolls back the transaction if commit isn't called.

I see TransactionManager trait and AnsiTransactionManager, and it looks like this is meant to be an internal impl detail of Connection, but it also seems like the most promising way to do this sort of procedural transaction invocation.  Am I on the right track?  Has anyone already done what I'm describing?

**Comment by weiznich** [comment]
We do not really support any other variant to execute transactions via our public API as there is too much that could go wrong with the other approaches. Using a guard object would cause issues if the rollback fails due to whatever reasons as we cannot return any result from the destructor.
`TransactionManager` is more or less considered to be a building block for third party connection implementations rather than something that should be used in ordinary applications.

That all written I would be interested to know in which cases you feel that the current `Connection::transaction` function does not work well. Could you provide an concrete example for this?

**Reply by vdods** [reply]
Really what I'm trying to get at is running async code within a DB transaction.  I'm not exactly sure how to do that, because the transaction takes a FnOnce, which doesn't allow for async code.

I tried something to the effect of
```
connection.transaction(|| {
    tokio::runtime::Handle::current().block_on(async {
        <body of transaction, including async calls>
    })
})
```
but this fails at runtime with `Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks`.

I imagine plenty of people have had this problem.  Is there a straightforward way to handle this?

**Reply by vdods** [reply]
This runs, but I don't think blocking is the right way to go.
```
connection.transaction(|| {
    let handle = tokio::runtime::Handle::current();
    let _ = handle.enter();
    futures::executor::block_on(async {
        <body of transaction>
    })
})            
```

**Reply by weiznich** [reply]
In my opinion it is an anti-pattern to perform IO other than database stuff inside of an transaction, as this will extend the time the transaction is open for a considerable amount of time. Async function calls in rust are most likely coupled with IO related operations, therefore they are not supported in this position.

**Reply by vdods** [reply]
Yeah, I agree.  But unfortunately my application requires some amount of synchronization between two databases across a network API, with a number of constraints on the "local" DB, so for now transactions it is.

I solved the problem by implementing my own transaction guard object which begins/commits/rolls back the transaction via raw sql.  Also not ideal, in that rollback errors are not caught, but in my case I throw away the connection after a transaction succeeds or fails anyway.

**Reply by weiznich** [reply]
Instead of using raw sql, you should likely use at least the existing `TransactionManger` interface so that transaction depth in tracked at connection level. That interface is part of the general public API of diesel, but we do not guarantee that a certain Connection implementation will continue to use a particular `TransactionManager` implementation forever.

**Reply by Alvenix** [reply]
seaorm crate provide this feature using a guard object. I read their drop code and their guard object panics on failure in the drop trait.

For me I see that trade off as worth it for the ergonomic wins but I understand diesel focus on correctness.

However, I note that I am probably not using transactions as much as actually needed for correctness for the applications I work on. Because I find using transactions make code less readable to people I work with.

Assuming I am using a framework which handles panics and a pooling crate, could the transaction error be handled automatically in these cases?

**Reply by weiznich** [reply]
> Because I find using transactions make code less readable to people I work with. 

Can you explain that a bit more, because in my opinion that's exactly the other way around: Needing to keep track of all early returns (by using `?` for example) when using a guard object is much harder to do correctly.

> Assuming I am using a framework which handles panics and a pooling crate, could the transaction error be handled automatically in these cases?

The diesel provided pooling integration with r2d2 currently considers a connection as broken if the connection is returned to a pool via panic or if it contains an open transaction. r2d2 will close such connections in both cases, which also drops open transactions without committing them. But that's both a last line of defense thing and should be avoided if possible as closing and reopening connections is quite expensive.

**Reply by Alvenix** [reply]
> But that's both a last line of defense thing and should be avoided if possible as closing and reopening connections is quite expensive.

Personally I would love to have the option for panicking guards and would probably be using them if they improve prototyping and readability. I always wanted to try seaorm for this feature along with few others. However, diesel is more popular and is stable.

> Can you explain that a bit more, because in my opinion that's exactly the other way around: Needing to keep track of all early returns (by using ? for example) when using a guard object is much harder to do correctly.

I am not sure why do we need to track early returns? I would assume that early return = some error happened = re-roll the transaction to be in more consistent state. 

Some of the people I work with are relatively new to Rust (and are still learning), which make reading and writing complex closures harder to them. I don't know how to quantify it, but I get asked really a lot about closures and their confusing errors. I know maybe we shouldn't optimize for these cases. Maybe I am overly cautious to complex closures and if we simply use them more liberally it would become easier for the team as we can simply copy the transaction closure signature around.

We used to have to write transactions like this to have them compile.  (anyhow here is for illustration only)

```
    conn.transaction::<_, anyhow::Error, _>(|conn| {});
```

But I refactored the code (while writing the comment) to remove the generic part and it worked. (The code changed a lot from before, I don't know why it works now)

```
conn.transaction(|conn| {});
```

I still think that closures are harder for beginners, because they frequently provide harder errors to debug when things go wrong (due to inference or variable capturing maybe). I don't have examples in mind but maybe if someone needed to use non-database logic which have a different error type in the closure.

Although unrelated using diesel through Rocket require closures. To access the connection from the pool we would use closures and nesting closures which also add complexity.

```
conn.run(|conn| {
    conn.transaction(|conn| {});
}).await;
```

**Reply by weiznich** [reply]
> I am not sure why do we need to track early returns? I would assume that early return = some error happened = re-roll the transaction to be in more consistent state.

Let's consider the following example:

```rust
let mut conn = establish_connection()?;

let transaction_guard = conn.begin_transaction();

let u = users::table.load(&mut conn)?;

let users_for_insert = process_users(u)?;

diesel::insert_into(users::table).values(&users_for_insert).execute(&mut conn)?;

transaction_guard.commit(&mut conn)?;

```

So based on your previous description that would cause a panic as soon as either the `load`, the `process_user` or the `execute` function call returned an error, because that will drop the transaction guard that contains an uncommitted transaction. Now you could argue that you would need to rollback the transaction on drop instead of panicing, but that raises concerns about error handling for sync code. For async code that's just not possible because you cannot perform async actions in a sync drop impl (that' likely why SeaORM panics in that case).

> We used to have to write transactions like this to have them compile. (anyhow here is for illustration only)
> ```
>    conn.transaction::<_, anyhow::Error, _>(|conn| {});
> ```

That's required in situations where rustc cannot infer a single error type. So mostly when you only use `?` and `Ok()` in the relevant callback.


> Although unrelated using diesel through Rocket require closures. To access the connection from the pool we would use closures and nesting closures which also add complexity.
> ```
> conn.run(|conn| {
>    conn.transaction(|conn| {});
> }).await;
> ```

There is a simple solution here: Use functions instead of callbacks, because a function also implements the necessary traits to be used in that location. For example:

```rust
conn.run(|conn| conn.transaction(function_run_in_transaction)).await

fn function_run_in_transaction(conn: &mut PgConnection) -> Result<Whatever, anyhow::Error> {
    // whatever
}
```

That also addresses your issue with requiring to specify the error type as that is now explicitly given anyway. 


That all written: It's totally fine to disagree with that written above and there is no need to wait for diesel to support these kind of transaction guard objects. You can easily write that on your own in a few lines of code:

```rust
trait TransactionExt {
    fn begin_transaction(&mut self) -> QueryResult<TransactionGuard>;
}

impl TransactionExt for PgConnection {
    fn begin_transaction(&mut self) -> QueryResult<TransactionGuard> {
        AnsiTransactionManager::begin_transaction(self)?;
        Ok(TransactionGuard(true))
    }
}

struct TransactionGuard(bool);

impl TransactionGuard {
    fn commit(self, conn: &mut PgConnection) -> QueryResult<()> {
        AnsiTransactionManager::commit_transaction(conn)?;
        self.0 = false;
        Ok(())
    }
    
    fn rollback(self, conn: &mut PgConnection) -> QueryResult<()> {
        AnsiTransactionManager::rollback_transaction(conn)?;
        self.0 = false;
        Ok(())
    }
}

impl Drop for TransactionGuard {
    fn drop(&mut self) {
        if self.0 { panic!("Open transaction") }
    }
}

```

**Reply by Alvenix** [reply]
> So based on your previous description that would cause a panic as soon as either the load, the process_user or the execute function call returned an error, because that will drop the transaction guard that contains an uncommitted transaction. Now you could argue that you would need to rollback the transaction on drop instead of panicing, but that raises concerns about error handling for sync code. For async code that's just not possible because you cannot perform async actions in a sync drop impl (that' likely why SeaORM panics in that case).

I meant that seaorm attempt rollback and panic on rollback [failure](https://docs.rs/sea-orm/0.11.2/src/sea_orm/database/transaction.rs.html#293-297). Sorry for not being clear.

Thank you for the detailed response.

Both suggestions seem good, using function instead of closures is a good alternative that I did not think deeply about, it may simplify the code a lot. I may try to create transaction guard via extension trait later. I love that rust support adding method to existing structs via traits, I kinda forget that sometimes.

**Reply by ColonelThirtyTwo** [reply]
I'll point out that one advantage to guard objects over a closure is that closures interfere with flow control. Using `break`, `continue`, and `return` are natural with a guard object, but require workarounds with a closure. And it's impossible to `await` inside of the closure for diesel's transactions (not that awaiting with a transaction open is a great idea, but there may be a use case for it).

**Reply by weiznich** [reply]
> I'll point out that one advantage to guard objects over a closure is that closures interfere with flow control. Using break, continue, and return are natural with a guard object, but require workarounds with a closure. And it's impossible to await inside of the closure for diesel's transactions (not that awaiting with a transaction open is a great idea, but there may be a use case for it).

At least for me that's more an advantage of the closure based approach. In all listed cases you either want to be explicit about what happens with your transaction (`break`, `continue`, `return`) or you want to prevent that at all (`await`). A drop guard based approach cannot guarantee that.

**Reply by cylewitruk** [reply]
I can chime in and give one useful example, anyway - I'm trying to retrofit Diesel into a large, older codebase which uses `rusqlite` and a lot of tx handles passed around in their own handle objects which are used to synchronize commit/rollback between multiple types of datastores (it's a blockchain, so the on-disk merkleized representation and the metadata stored in sqlite).

Having an opt-in "procedural-tx" feature or something that would expose the `TransactionManager` from a `Connection` would have been really helpful here, as the changes become much more extensive and risky when also having to try to change the functionality of the code at the same time as trying to convert to Diesel.

I found https://github.com/DR-BoneZ/diesel-transaction-handles-rs which is a nice idea which I might adapt, but I would have preferred to avoid the overhead :)

**Reply by weiznich** [reply]
> Having an opt-in "procedural-tx" feature or something that would expose the TransactionManager from a Connection would have been really helpful here, as the changes become much more extensive and risky when also having to try to change the functionality of the code at the same time as trying to convert to Diesel.

That's already possible as the `TransactionManager` type is already exposed: https://docs.diesel.rs/2.1.x/diesel/connection/trait.Connection.html#associatedtype.TransactionManager. You can also access the actual transaction manager implementation via `AnsiTransactionManager`: https://docs.diesel.rs/2.1.x/diesel/connection/struct.AnsiTransactionManager.html
